### ISSUE TITLE
Document build customization to disable analyzers when building in Visual Studio

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#1711]
Some users are interested in disabling analyzers in Visual Studio to improve build performance. Specifically:

* Enable analyzers for IntelliSense purposes
* Enable analyzers for command line builds
* Disable analyzers when actually building the project inside of Visual Studio

The documentation for this functionality should be preceded by a clear notice regarding the potential downside of this behavior.

> :warning: Using this build customization may change the build behavior in ways that increase the chances of developers submitting code which does not compile in the automated build environment. In particular, projects with any of the following characteristics are advised to avoid the use of this:
>
> * Projects with `/warnaserror` enabled
> * Projects with one or more analyzers installed that have a default severity of Error
> * Projects which use a rule set file to set the severity of one or more analyzers to Error

This can be accomplished by adding the following to a project file:

```xml
<Target Name="DisableAnalyzersForVisualStudioBuild"
        BeforeTargets="CoreCompile"
        Condition="'$(BuildingInsideVisualStudio)' == 'True' And '$(BuildingProject)' == 'True'">
  <!--
    Disable analyzers when building a project inside Visual Studio. Note that analyzer behavior for IntelliSense
    purposes is not altered by this.
  -->
  <ItemGroup>
    <Analyzer Remove="@(Analyzer)"/>
  </ItemGroup>
</Target>
```

**Comment by whut** [comment]
:+1:

**Comment by felickz** [comment]
Have you considered folloing the CA setting treating warnings as errors?

`<CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>`

I find this setting helpful where i want to show warnings to dev's on localhost, but Errors in a CI build..

```
  <PropertyGroup Condition=" '$(TF_BUILD)' == 'True' ">
	<CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
  </PropertyGroup>
```

**Comment by sharwell** [comment]
@felickz We don't have access to that from inside an analyzer, so I'm not sure there's anything we can do to support it directly.

**Comment by Trass3r** [comment]
btw Resharper Build now sets 'BuildingByReSharper' just like VS sets BuildingInsideVisualStudio.

**Comment by claudiopi** [comment]
@sharwell the workaround was working until VS Update 2 but doesn't seem to work anymore in VS Update 3. Are you aware of any changes or an easier way in order to disable Analyzers when building inside Visual Studio ? 

Thanks!

**Comment by ankitvijay** [comment]
HI @claudiopi did you you get way to disable Analyzer inside VS 2015 Update 3? We have the same issue here.

**Comment by onyxmaster** [comment]
@sharwell Unfortunately this doesn't work in VS 2017 15.3 with "new" csproj format. The condition evaluates as true even for "IntelliSense" phase, effectively removing analyzers completely even from project view.

An example of project file for VS2017 15.3:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="StyleCop.Analyzers" Version="1.1.0-beta004" />
  </ItemGroup>

  <Target Name="DisableAnalyzersForVisualStudioBuild" BeforeTargets="CoreCompile" Condition="'$(BuildingInsideVisualStudio)' == 'True' And '$(BuildingProject)' == 'True'">
    <ItemGroup>
      <Analyzer Remove="@(Analyzer)"/>
    </ItemGroup>
  </Target>
</Project>
```

**Comment by onyxmaster** [comment]
I was able to "fix" the problem by using adding the `And '$(BuildProjectReferences)' == 'True'` to condition, but I'm not sure this doesn't have any unintended side effects.

**Comment by sharwell** [comment]
@onyxmaster Thanks for the heads-up. Can you try the following?

1. First, explicitly import the Sdk.props and Sdk.targets files:

    ```xml
    <Project> <!-- No 'Sdk' attribute -->
      <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.props" /> <!-- First line -->

      <!-- ... -->

      <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.targets" /> <!-- Last line -->
    </Project>
    ```

2. Then, move the `<Target>` definition from far above *after* the import of **Sdk.targets** (at the end of the project).

3. Modify the condition by replacing `'$(BuildingProject)' == 'true'` with the following:

    ```
    '$(DesignTimeBuild)' != 'true' And '$(BuildingProject)' == 'true'
    ```

**Comment by sharwell** [comment]
💭 Using the above approach, it may actually be possible to eliminate the `<Target>` altogether and simply use an `<ItemGroup>`, but I haven't investigated this.

```xml
<!-- ItemGroup must be placed after explicit Sdk.targets import -->
<ItemGroup Condition="'$(BuildingInsideVisualStudio)' == 'True' And '$(DesignTimeBuild)' != 'True' And '$(BuildingProject)' == 'True'">
  <Analyzer Remove="@(Analyzer)" />
</ItemGroup>
```

**Comment by onyxmaster** [comment]
Thank you, I'll try this when I finish upgrading our projects.

**Comment by glenn-slayden** [comment]
Another approach is to `#ifdef`-out **C#** source code that you don't want Intellisense (etc.) to see. The following currently works for that purpose:

    <PropertyGroup>
        <DefineConstants Condition="'$(BuildingProject)' == 'false'">ANALYZER;$(DefineConstants)</DefineConstants>
    </PropertyGroup>

With this defined, you can do things like the following example, which provides the required Intellisense stub for an otherwise-stock use of a **.netmodule** source file getting compiled into a standard MsBuild **C#** project. 

	#if ANALYZER
	public static class ABCDE
	{
		[MethodImpl(MethodImplOptions.ForwardRef)]
		public static extern String Test();
	}
	#endif

The MsBuild "item" entry for providing the implementation behind this `MethodImplOptions.ForwardRef` would be, e.g.:

    <ItemGroup>
        <AddModules Include="ABCDE.netmodule" />
    </ItemGroup>

Fwiw, the **.netmodule** itself can also be built using with a fairly unadulterated MsBuild csproj with type using `<OutputType>Module</OutputType>`. It's critical for this to set `NoWin32Manifest` to 'true'. Better yet, change that project tag to `<Project DefaultTargets="Compile" ...` and pick up the **.netmodule** from the `\obj` directory.

**Comment by Livven** [comment]
@glenn-slayden Coincidence indeed! Wouldn't your approach require surrounding every file with those ifdefs and also break the build (since nothing gets compiled then) or am I misunderstanding something?

**Comment by glenn-slayden** [comment]
Yes, it would. My note is not as well suited if you're looking for ways to improve build performance in bulk.

**Comment by SeidChr** [comment]
Ist there currently anything "out of the box" i can set on build-level (wihtout modifying the 30+ project files where we use stylecop) to disable the analyzers for this build completely?
We need this in visual studio builds and tfs builds. Only thing we want is Highlighting.

**Comment by sharwell** [comment]
@SeidChr I'm not sure which version of the compiler tools first supports it, but you can pass `/p:RunAnalyzersDuringBuild=false` to the build to turn them off for the command line.

**Comment by SeidChr** [comment]
Perfect, thank you!

**Comment by hhyyrylainen** [comment]
Is there a way to make this work with mono? I've just about tried everything I can, but I just can't get builds working so that I would have a configuration that would allow me faster debugging by not needing to run the analysis every time.

**Comment by springy76** [comment]
@sharwell regarding your OP

>     <Analyzer Remove="@(Analyzer)"/>

Of course some of my colleges adopted your code years ago to shorten CI-build logs (too much stylecop warnings) and this week I've spent hours of googling and analyzing why after switching to net6 all CI-builds don't have working razor pages (but everything else like static content, swagger, API works) - effectively this special project only has 1 razor page with dev info which isn't essential for the app but I still wondered why it is missing now.

Guess what, I think you already know: Since NET5 there are more kinds of "Analyzers" which are faaaaaaaar moooooooore than just a simple analyzer - but some poor design(er) mixed them up being exactly at the same position:
 * System.Text.Json.SourceGeneration
 * Microsoft.Extensions.Logging.Generators
 
and finally starting with NET6
 * Microsoft.NET.Sdk.Razor.SourceGenerators

So my "analyzers" node in VS today consists of 7 real analyzers and 4 source generators, 9 of them default in a aspnet project +2 of stylecop. The word "generator" or "analyzer" at end of line, many times I have to scroll to see which one is which: Frustrating.

**Comment by sharwell** [comment]
@springy76 With newer versions of the compiler you should be using the `RunAnalyzersDuringBuild` and related properties:
[Disable source code analysis for .NET (docs.microsoft.com)](https://docs.microsoft.com/en-us/visualstudio/code-quality/disable-code-analysis?view=vs-2022#net-framework-projects)
